### ISSUE TITLE
Stage 1 - training losses - todo contrastive_loss

### ISSUE THREAD

**Issue by johndpope** [johndpope/MegaPortrait-hack#14]
https://github.com/johndpope/MegaPortrait-hack/blob/main/train.py#L196

adding any of the others hits a bug with gradients.

**Comment by Jie-zju** [comment]
Have you noticed  "losses are calculated using only foreground regions" mentioned in Paper? I tried implement the network (include somewhat different from paper ) and train it. However,  just use loss on all images for training seems fail, since the training stage is not stable.  I think  "losses are calculated using only foreground regions" May Be Important ! Please attention on it!

**Comment by johndpope** [comment]
thanks for pointing this out. I fix later.

**Comment by johndpope** [comment]
i draft this - yet to test it out.
https://github.com/johndpope/MegaPortrait-hack/pull/21/files

so much flux - needs stabilising. 
fyi @robinchm  


I look at the patchgan stuff

**Comment by johndpope** [comment]
warped_driving_frame:torch.Size([3, 129, 129])

on my training branch - i implement and fix - but now the warped / cropped image is small....
i can make it bigger - or I have to work out how the network can accept smaller images....
not sure.
 https://github.com/johndpope/MegaPortrait-hack/tree/feat/14-training

**Comment by johndpope** [comment]
![Screenshot from 2024-05-30 12-15-54](https://github.com/johndpope/MegaPortrait-hack/assets/289994/69aa8329-f99c-4056-bfe1-682696158f0b)

some progress.
